getopts rewrite of AAXtoMP3, merges, --targetdir, Readme updates.

upuv opened the pull request:
### Comments 
Complete rewrite of options processing.   Much more flexible for future changes.  Now supports both short and long option formats.  Also options are now unordered.

### Changes:
* Merged in the -aac -opus encodings
* Created a --targetdir option to all for output to go to another directory.
* Lots of Readme.md updates
* Lots of variable validation code added
* Changed the .authcode handling again.  It's now a option as it is optional on the command line.  I've also added support for local and home directory .authcode support. 

### Outstanding Issues:
flac file does not want to split into multiple parts in my testing so I adjusted the option to force single for the time being.

upuv commented:
### Comments ###
Primary changes revolve around DEBUG output.  We now have a functional -d --debug option.  The effort is to enable others to enhance the code a little more cleanly.

### Changed ###
* added debug_file function now dumps the contents of a text file to output
* modified debug function to only output if -d or --debug flag is set.
* added log function to replace old debug function
* moved variable computation in ffmpeg commands to resolve in variable prior to ffmpeg.
* removed AAXtoMPA as it is redundant now that --aac exists
* added various comments in code.
* updated Readme.md

### Issues ###
*  --flac option will only output a single file.  The split file throughs an error.
* Specific container options are something that should be explored.

jostyee commented:
@upuv Can you do a rebase? thank you.

upuv commented:
### Apologize ###
Sorry about that.  My bad. Sorry for the lack of rebase.  &lt;hand slap&gt;

### Comments. ###
I recommend testing on your platform with the -d and and short form options for most things.  Please refer to the Readme.md for details.  The -d or --debug options are extremely verbose.

Also note I have create a code structure the makes options un ordered.  Removing the previous MUST BE IN ORDER requirement.

jostyee commented:
I'll test it in the following days, Thanks again.

upuv commented:
## Changes ##
I added m4b support to the mp3 script
I removed the m4b script.
I updated Readme.md file.

jostyee commented:
Great, I was about to remove the m4b script myself, following the upstream update.
I'll take more time to test thoroughly.

jostyee commented:
hey @upuv , the intention of my fork was to add artwork cover to the m4b files converted, for which I did:

```shell
brew install mp4v2
```

then invoke it at https://github.com/jostyee/AAXtoM4B/blob/master/AAXtoM4B#L108

Since you removed this line, do you have a better alternative?

upuv commented:
Good Spot that was a genuine regression on my part.

I will fix this and add check code to make sure the command is present.  I will not fail mp4 creation I will only WARN if command not present.

While coding up the fix I figured out how I managed to miss that.  What I was doing was diffing AAXtoM4A and AAXtoM4B to see the difference.  When I merged in --aac code earlier that took care of creating m4a files, but it left out the cover art.  

All good in the end.  I also added appropriate command verification which no one had previously for mp4art.

upuv commented:
### Changes ###
* m4a and m4b now have cover art added to output file.  Fixing a regresion I introduced.
* m4a and m4b better documented in Readme.
* cover art inclusion will only occur if mp4art is installed.  If not the process will still succeed just with out the cover added to the output.
* added install instructions to code for mp4art on MacOS and Ubuntu.  Probably should move this to Readme.

jostyee commented on AAXtoMP3:
```diff
@@ -1,166 +1,305 @@
+                      # Identical to --acc option.
+    -e:m4a            ) codec=copy; extension=m4a; mode=single;                     shift ;;
+                      # Similiar to --aac but specific to audio books.
+    -e:m4b            ) codec=copy; extension=m4a; mode=single; container=m4b       shift ;;
```
Can you unset `mode=single` in this param? I think m4b is better with split files.

upuv replied:
How about this.  I add a --chaptered switch to force split files.

The original design by Apple was that a single m4b file contains all of the chapters in one file. Since the m4b already internally has chapter markers.

Which I added in the next code version.  There will be conditions on this documented in the readme.  As this flag will break --flac selection.  Note flac splitting is a known bug in ffmpeg.

jostyee replied:
Sure, it'd be a better solution, thanks.

upuv replied:
This actually took a little more work than originally planned. 

I had to extend the concept of container in the code.  So that got salt and peppered everywhere.
Also the defaults had to be better documented in the README.

It also involved a lot more error checking.

As part of the work I also cleaned up the logging a bit more.  I made the splitting less verbose by silencing ffmpeg output.  Instead I put a human readable set of logs.  I also added cover file to the m4a m4b chapter files.

jostyee commented on AAXtoMP3:
```diff
@@ -1,166 +1,305 @@
-    shift
+# -----
+# Check the target dir for if set if it is writable
+if [[ ! -w "${targetdir}" || ! -d "${targetdir}" ]] ; then
```
This check should be after https://github.com/jostyee/AAXtoM4B/pull/3/files#diff-f1655c3b19357414d6913f2a6bdc7f37R197 , which set a default output directory, otherwise when don't pass `-t` the script would fail.

upuv replied:
Yep there is a bug there.   I added a check to see if the -t option was passed.  If it was then it checks for the target directory existence and is writable.

upuv commented:
### Changes
* Added --chaptered option
* Chaptered m4a and m4b support added.  Added cover file to each chapter
* Fixed --target_dir bug
* Modified log out.   Chaptering output now human readable
* Modified chaptered file naming removed excess whitespace
* Updated README.md

### Tag
* upuv_v2.1.0

jostyee commented on AAXtoMP3:
```diff
@@ -1,166 +1,325 @@
-    exit 1
-fi
+# Usage Synopsis.
+usage=$'\nUsage: AAXtoMP3.sh [--flac] [--aac] [--opus ] [--single] [--chaptered]\n[-e:m4a] [-e:m4b] [--authcode <AUTHCODE>] [--output_dir <PATH>] {FILES}\n'
```
`AAXtoMP3.sh` -> `AAXtoMP3`

upuv replied:
Change uploading now.

jostyee commented on AAXtoMP3:
```diff
@@ -1,166 +1,325 @@
+  # If we want multiple file we take the big mp3 and split it by chapter.
+  # Not all audio encodings make sense with multiple chapter outputs.  See options section
+  # for more detail
+  if [ "${mode}" == "chaptered" ]; then
```
We need to check `${extension} == "mp3`, or `AAXtoMP3 -e:m4b -c` would generate useless m3u file, too.

upuv replied:
Actually it's not useless.  The playlist is still valid and works with some players.

However there was a couple of missing ; in the options.  Fixing and retesting now.

jostyee replied:
Tried one and got:
```
#EXTM3U
#EXTINF:4154,Flowers for Algernon - Chapter 1
Flowers for Algernon-1 Chapter 1.mp3
#EXTINF:4126,Flowers for Algernon - Chapter 2
Flowers for Algernon-2 Chapter 2.mp3
#EXTINF:3859,Flowers for Algernon - Chapter 3
Flowers for Algernon-3 Chapter 3.mp3
#EXTINF:4216,Flowers for Algernon - Chapter 4
Flowers for Algernon-4 Chapter 4.mp3
#EXTINF:4288,Flowers for Algernon - Chapter 5
Flowers for Algernon-5 Chapter 5.mp3
#EXTINF:3687,Flowers for Algernon - Chapter 6
Flowers for Algernon-6 Chapter 6.mp3
#EXTINF:3969,Flowers for Algernon - Chapter 7
Flowers for Algernon-7 Chapter 7.mp3
#EXTINF:4033,Flowers for Algernon - Chapter 8
Flowers for Algernon-8 Chapter 8.mp3
```

Since the extension is `m4b` not `mp3` , load the m3u would complain files are missing.

upuv commented:
### Changes
* Fixed container setting in options.  Missing ; 
* Fixed Usage String.

Thanks jostyee for all the testing.

jostyee commented:
Bravo, Thank you @upuv for contributing.
